ipn/ipnserver/server.go

```diff
@@ -63,22 +64,47 @@ func (s *Server) mustBackend() *ipnlocal.LocalBackend {
 	return lb
 }
 
+// waiterSet is a set of callers waiting on something. Each item (map value) in
+// the set is a func that wakes up that waiter's context. The waiter is responsible
+// for removing itself from the set when woken up. The (*waiterSet).add method
+// returns a cleanup method which does that removal. The caller than defers that
+// cleanup.
+//
+// TODO(bradfitz): this is a generally useful pattern. Move elsewhere?
+type waiterSet set.HandleSet[context.CancelFunc]
```

Comment:
if the goal is to wake up everyone on an event, you could also do something like this and not have to worry about all the accounting and cleanup.

```
type waiterSet struct {
  ch syncs.AtomicValue[chan struct{}]
}
func (s *waiterSet) get() (ready <-chan struct{}) {
  return s.ch.Load()
}
func (s *waiterSet) wakeAll() {
  close(s.ch.Swap(make(chan struct{})))
}
```

Reply:
you will need to initialize it once

Reply:
^ I've likewise had good luck using this simpler pattern for the broadcast case (though I did it with explicit locks, https://pkg.go.dev/github.com/creachadair/msync#Trigger).

Reply:
Per chat, we're going to do this later in its own PR adding the type/docs/tests and convert multiple callers.